plugins/out_stackdriver/stackdriver_operation.c

```diff
@@ -64,81 +69,82 @@ int extract_operation(flb_sds_t *operation_id, flb_sds_t *operation_producer,
-                        if (tmp_p->val.type != MSGPACK_OBJECT_BOOLEAN) {
-                            continue;
-                        }
-                        if (tmp_p->val.via.boolean) {
-                            *operation_last = FLB_TRUE;
-                        }
-                    }
-                    else {
-                        /* extra sub-fields */ 
-                        *extra_subfields += 1;
-                    }
-
-                }
+        op_status = OPERATION_EXISTED;
+        msgpack_object sub_field = p->val;
+        
+        tmp_p = sub_field.via.map.ptr;
+        tmp_pend = sub_field.via.map.ptr + sub_field.via.map.size;
+
+        /* Validate the subfields of operation */
+        for (; tmp_p < tmp_pend; ++tmp_p) {
+            if (tmp_p->key.type != MSGPACK_OBJECT_STR) {
+                continue;
+            }
+
+            if (assign_subfield_str(tmp_p, OPERATION_ID, 
+                                    OPERATION_ID_SIZE, operation_id) == FLB_TRUE) {
+                continue;
+            }
+            else if (assign_subfield_str(tmp_p, OPERATION_PRODUCER, 
+                                         OPERATION_PRODUCER_SIZE, 
+                                         operation_producer) == FLB_TRUE) {
+                continue;
+            }
+            else if (assign_subfield_bool(tmp_p, OPERATION_FIRST, 
+                                          OPERATION_FIRST_SIZE, 
+                                          operation_first) == FLB_TRUE) {
+                continue;
+            }
+            else if (assign_subfield_bool(tmp_p, OPERATION_LAST, 
+                                          OPERATION_LAST_SIZE, 
+                                          operation_last) == FLB_TRUE) {
+                continue;
+            }
+            else {
+                *extra_subfields += 1;
             }
         }
     }
     
     return op_status == OPERATION_EXISTED;
 }
 
-void pack_extra_operation_subfields(msgpack_packer *mp_pck, msgpack_object *operation, int extra_subfields) {
+void pack_extra_operation_subfields(msgpack_packer *mp_pck, 
+                                    msgpack_object *operation, int extra_subfields) {
     msgpack_object_kv *p = operation->via.map.ptr;
     msgpack_object_kv *const pend = operation->via.map.ptr + operation->via.map.size;
 
     msgpack_pack_map(mp_pck, extra_subfields);
 
     for (; p < pend; ++p) {
-        if (strncmp("id", p->key.via.str.ptr, p->key.via.str.size) != 0 
-            && strncmp("producer", p->key.via.str.ptr, p->key.via.str.size) != 0
-            && strncmp("first", p->key.via.str.ptr, p->key.via.str.size) != 0
-            && strncmp("last", p->key.via.str.ptr, p->key.via.str.size) != 0) {
-            msgpack_pack_object(mp_pck, p->key);
-            msgpack_pack_object(mp_pck, p->val);
+        if (cmp_str_helper(p->key, OPERATION_ID, OPERATION_ID_SIZE)
```

Comment:
is it convention that we don't need to check whether FLB boolean is true or not? If yes, I am fine with this.

Reply:
Yes